README.md
# prometheus_exporters

Cookbook to install and configure various Prometheus exporters on systems to be monitored by Prometheus.

Currently supported exporters are node, postgres, redis, mysqld, haproxy, process, apache, blackbox, snmp, statsd, and wmi. More may be added in the future. Please contact the author if you have specific requests.

All of the exporters are available as chef custom resources that can be instantiated from other cookbooks.


# Supports

* Ubuntu 16.04
* Ubuntu 18.04
* Fedora 31
* Fedora 30
* Debian 4
* Debian 9
* CentOS 7
* Amazon Linux

And probably other RHEL or Debian based distributions.

* Windows Server 2012 & 2016 (wmi\_exporter recipe only)

Tests are made using last available Chef 15 along with latest Chef 14.

# Resource List

- [apache_exporter](https://github.com/evilmartians/chef-prometheus-exporters#apache_exporter)
- [blackbox_exporter](https://github.com/evilmartians/chef-prometheus-exporters#blackbox_exporter)
- [consul_exporter](https://github.com/evilmartians/chef-prometheus-exporters#consul_exporter)
- [elasticsearch_exporter](https://github.com/evilmartians/chef-prometheus-exporters#elasticsearch_exporter)
- [haproxy_exporter](https://github.com/evilmartians/chef-prometheus-exporters#haproxy_exporter)
- [mongodb_exporter](https://github.com/evilmartians/chef-prometheus-exporters#mongodb_exporter)
- [mysqld_exporter](https://github.com/evilmartians/chef-prometheus-exporters#mysqld_exporter)
- [node_exporter](https://github.com/evilmartians/chef-prometheus-exporters#node_exporter)
- [postgres_exporter](https://github.com/evilmartians/chef-prometheus-exporters#postgres_exporter)
- [process_exporter](https://github.com/evilmartians/chef-prometheus-exporters#process_exporter)
- [rabbitmq_exporter](https://github.com/evilmartians/chef-prometheus-exporters#rabbitmq_exporter)
- [redis_exporter](https://github.com/evilmartians/chef-prometheus-exporters#redis_exporter)
- [snmp_exporter](https://github.com/evilmartians/chef-prometheus-exporters#snmp_exporter)
- [statsd_exporter](https://github.com/evilmartians/chef-prometheus-exporters#statsd_exporter)
- [varnish_exporter](https://github.com/evilmartians/chef-prometheus-exporters#varnish_exporter)
- [wmi_exporter](https://github.com/evilmartians/chef-prometheus-exporters#wmi_exporter)

# Resources

## apache_exporter

* `insecure` Ignore server certificate if using https. (default false)
* `log_format` Output format of log messages. e.g. "logger:syslog?appname=bob&local=7" or "logger:stdout?json=true" (default: "logger:stdout")
* `log_level` The logging level. (default: "info")
* `scrape_uri` URI to apache stub status page. (default "http://localhost/server-status/?auto")
* `telemetry_address` Address on which to expose metrics. (default ":9117")
* `telemetry_endpoint` Path under which to expose metrics. (default "/metrics")
* `user` User under whom to start apache exporter. (default: "root")

```ruby
apache_exporter 'main' do
  scrape_uri "http://localhost:8090/server-status/?auto"
  telemetry_address ":9118"
  telemetry_endpoint "/_metrics"
end
```

## blackbox_exporter

This exporter requires a config file. Read more [here](https://github.com/prometheus/blackbox_exporter/blob/master/CONFIGURATION.md). For basic usage the default `blackbox.yml` should be sufficient.

* `config_file` default: `/opt/blackbox_exporter-#{node['prometheus_exporters']['blackbox']['version']}.linux-amd64/blackbox.yml`
* `log_level` Only log messages with the given severity or above. Valid levels: [debug, info, warn, error]
* `timeout_offset` default: 0.5 Offset to subtract from timeout in seconds.
* `user` User under whom to start blackbox exporter. (default: "root")
* `web_listen_address` Address to listen on for web interface and telemetry. (default: ":9115")

```ruby
blackbox_exporter 'main'
```

## consul_exporter

* `consul_allow_stale` Allows any Consul server (non-leader) to service a read. (default: false)
* `consul_ca_file` File path to a PEM-encoded certificate authority used to validate the authenticity of a server certificate.
* `consul_cert_file` File path to a PEM-encoded certificate used with the private key to verify the exporter's authenticity.
* `consul_health_summary` Collects and exports information about each registered service. (default: false)
* `consul_insecure` Disable TLS host verification. (default: false)
* `consul_key_file` File path to a PEM-encoded private key used with the certificate to verify the exporter's authenticity.
* `consul_require_consistent` Enforce fully consistency on Consul reads. (default: false)
* `consul_server_name` Override the hostname for the TLS certificate. It can be used to ensure that the certificate name matches the decalred hostname.
* `consul_server` Address of the Consul instance to connect to. (default: "http://localhost:8500")
* `consul_timeout` Timeout on HTTP requests to consul. (default: "500ms")
* `kv_prefix` Prefix from which to expose key/value pairs.
* `kv_filter` Regex that determines which keys to expose. (default: ".*")
* `log_format` Output format of log messages. One of: [logfmt, json]. (default: "logfmt")
* `log_level` The logging level. (default: "info")
* `user` User under whom to start elasticsearch exporter. (default: "root")
* `web_listen_address` Address to listen on for web interface and telemetry. (default: ":9107")
* `web_telemetry_path` Path under which to expose metrics. (default: "/metrics")

```ruby
consul_exporter 'main' do
  action %i[install enable start]
end
```

## elasticsearch_exporter

* `es_all` If true, query stats for all nodes in the cluster, rather than just the node we connect to. (default: false)
* `es_ca` Path to PEM file that contains trusted Certificate Authorities for the Elasticsearch connection.
* `es_client_cert` Path to PEM file that contains the corresponding cert for the private key to connect to Elasticsearch.
* `es_client_private_key	` Path to PEM file that contains the private key for client auth when connecting to Elasticsearch.
* `es_cluster_settings` If true, query stats for cluster settings. (default: false)
* `es_clusterinfo_interval` Cluster info update interval for the cluster label. (default: 5m)
* `es_indices_settings` If true, query settings stats for all indices in the cluster. (default: false)
* `es_indices` If true, query stats for all indices in the cluster. (default: false)
* `es_shards` If true, query stats for all indices in the cluster, including shard-level stats (implies es.indices=true). (default: false)
* `es_snapshots` If true, query stats for the cluster snapshots. (default: false)
* `es_ssl_skip_verify` Skip SSL verification when connecting to Elasticsearch. (default: false)
* `es_timeout` Timeout for trying to get stats from Elasticsearch. (ex: 20s) (default: 5s)
* `es_uri` Address (host and port) of the Elasticsearch node we should connect to. (default: "http://localhost:9200")
* `log_format` Set the log target and format. Valid values are: [json, logfmt]
* `log_level` Only log messages with the given severity or above. Valid levels: [debug, info, warn, error, fatal]
* `user` User under whom to start elasticsearch exporter. (default: "root")
* `web_listen_address` Address to listen on for web interface and telemetry. (default: ":9114")
* `web_telemetry_path` Path under which to expose metrics. (default: "/metrics")

```ruby
elasticsearch_exporter 'main' do
  action %i[install enable start]
end
```

## haproxy_exporter

Monitor HAProxy metrics and stats. Read more [here](https://github.com/prometheus/haproxy_exporter).

* `haproxy_pid_file` Path to HAProxy pid file.
* `haproxy_scrape_uri` URI on which to scrape HAProxy.
* `haproxy_server_metric_fields` Comma-separated list of exported server metrics.
* `haproxy_ssl_verify` Flag that enables SSL certificate verification for the scrape URI.
* `haproxy_timeout` Timeout for trying to get stats from HAProxy.
* `log_format` Where to send log files. (default: "logger:stdout")
* `log_level` Only log messages with the given severity or above. Valid levels: [debug, info, warn, error, fatal]. (default: "info")
* `user` User under whom to start haproxy exporter. (default: "root")
* `web_listen_address` Address to listen on for web interface and telemetry. (default: "0.0.0.0:9116")
* `web_telemetry_path` Path under which to expose metrics. (default: "/metrics")

```ruby
haproxy_exporter 'main' do
  haproxy_scrape_uri 'http://user:pass@haproxy.example.com/haproxy?stats;csv'
end

haproxy_exporter 'main' do
  haproxy_scrape_uri 'unix:/run/haproxy/admin.sock'
  user 'haproxy'
end
```

Use the given defaults or set the attributes...

* `node['prometheus_exporters']['listen_interface']`
* `node['prometheus_exporters']['haproxy']['port']`
* `node['prometheus_exporters']['haproxy']['scrape_uri']`
* `node['prometheus_exporters']['haproxy']['ssl_verify']`
* `node['prometheus_exporters']['haproxy']['user']`

and add `recipe['prometheus_exporters::haproxy]` to your run_list.

## mongodb_exporter

* `collect_collection` Enable collection of Collection metrics
* `collect_connpoolstats` Collect MongoDB connpoolstats
* `collect_database` Enable collection of Database metrics
* `collect_indexusage` Enable collection of per index usage stats
* `collect_topmetrics` Enable collection of table top metrics
* `log_format` Set the log target and format. Example: "logger:syslog?appname=bob&local=7" or "logger:stdout?json=true"
* `log_level` Only log messages with the given severity or above. Valid levels: [debug, info, warn, error, fatal]
* `mongodb_authentication_database` Specifies the database in which the user is created
* `mongodb_max_connections` Max number of pooled connections to the database. (default: 1)
* `mongodb_socket_timeout` Amount of time to wait for a non-responding socket to the database before it is forcefully closed. Valid time units are 'ns', 'us' (or 'µs'), 'ms', 's', 'm', 'h'. (default: 3s)
* `mongodb_sync_timeout` Amount of time an operation with this session will wait before returning an error in case a connection to a usable server can't be established. (default: 1m)
* `mongodb_uri` MongoDB URI, format ([mongodb://][user:pass@]host1[:port1][,host2[:port2],...][/database][?options])
* `user` User under whom to start mongodb exporter. (default: "root")
* `web_auth_file` Path to YAML file with server_user, server_password keys for HTTP Basic authentication (overrides HTTP_AUTH environment variable).
* `web_listen_address` Address to listen on for web interface and telemetry. (default: ":9216")
* `web_ssl_cert_file` Path to SSL certificate file.
* `web_ssl_key_file` Path to SSL key file.
* `web_telemetry_path` Path under which to expose metrics. (default: "/metrics")

```ruby
mongodb_exporter 'main' do
  action %i[install enable start]
end
```

## mysqld_exporter

The mysqld_exporter resource supports running multiple copies of the MySQL exporter on the same system.

* `collector_flags` Specify which collector flags you wish to use. default: see belo
* `config_my_cnf` Path to .my.cnf file to read MySQL credentials from. (default: ~/.my.cnf)
* `data_source_name` MySQL connection string
* `instance_name` name of MySQL exporter instance. (**name attribute**)
* `log_format` If set use a syslog logger or JSON logging. Example: logger:syslog?appname=bob&local=7 or logger:stdout?json=true. Defaults to stderr.
* `log_level` Only log messages with the given severity or above. Valid levels: [debug, info, warn, error, fatal].
* `user` System user to run exporter as. (default "mysql")
* `web_listen_address` Address to listen on for web interface and telemetry. (default "127.0.0.1:9104")
* `web_telemetry_path` Path under which to expose metrics. (default "/metrics")

(default collector flags)

```
'\
-collect.global_status \
-collect.engine_innodb_status \
-collect.global_variables \
-collect.info_schema.clientstats \
-collect.info_schema.innodb_metrics \
-collect.info_schema.processlist \
-collect.info_schema.tables.databases \
-collect.info_schema.tablestats \
-collect.slave_status \
-collect.binlog_size \
-collect.perf_schema.tableiowaits \
-collect.perf_schema.indexiowaits \
-collect.perf_schema.tablelocks'
```

```ruby
mysqld_exporter 'main' do
  data_source_name '/'
  config_my_cnf '~/.my/cnf'
  user 'mysql'
end
```

## node_exporter

* `collector_diskstats_ignored_devices` Regexp of devices to ignore for diskstats. (default: "^(ram|loop|fd|(h|s|v|xv)d[a-z]|nvme\\d+n\\d+p)\\d+$")
* `collector_filesystem_ignored_fs_types` Regexp of filesystem types to ignore for filesystem collector. (default: "^(sys|proc|auto)fs$")
* `collector_filesystem_ignored_mount_points` Regexp of mount points to ignore for filesystem collector. (default: "^/(sys|proc|dev)($|/)")
* `collector_netclass_ignored_devices` Regexp of net devices to blacklist (mutually exclusive to device-whitelist) (default: "^$")
* `collector_netdev_device_blacklist` Regexp of net devices to ignore for netdev collector. (default: "^$")
* `collector_netdev_device_whitelist` Regexp of net devices to whitelist (mutually exclusive to device-blacklist) (default: "^$")
* `collector_ntp_ip_ttl` IP TTL to use while sending NTP query. (default: "1")
* `collector_ntp_local_offset_tolerance` Offset between local clock and local ntpd time to tolerate. (default: "1ms")
* `collector_ntp_max_distance` Max accumulated distance to the root. (default: "3.46608s")
* `collector_ntp_protocol_version` NTP protocol version. (default: "4")
* `collector_ntp_server_is_local` Certify that collector.ntp.server address is the same local host as this collector.
* `collector_ntp_server` NTP server to use for ntp collector. (default: "127.0.0.1")
* `collector_qdisc_fixtures` Test fixtures to use for qdisc collector end-to-end testing.
* `collector_perf_cpus` specify a list of alternate CPUs.
* `collector_runit_servicedir` Path to runit service directory.
* `collector_supervisord_url` XML RPC endpoint. (default: "http://localhost:9001/RPC2")
* `collector_systemd_enable_restarts_metrics` Enables service unit metric service\_restart\_total
* `collector_systemd_enable_start_time_metrics` Enables service unit metric unit\_start\_time\_seconds
* `collector_systemd_enable_task_metrics` Enables service unit tasks metrics unit\_tasks\_current and unit\_tasks\_max
* `collector_systemd_private` Establish a private, direct connection to systemd without dbus.
* `collector_systemd_unit_blacklist` Regexp of systemd units to blacklist. Units must both match whitelist and not match blacklist to be included. (default: ".+\\.(automount|device|mount|scope|slice)")
* `collector_systemd_unit_whitelist` Regexp of systemd units to whitelist. Units must both match whitelist and not match blacklist to be included. (defaut: ".+")
* `collector_textfile_directory` Directory to read text files with metrics from. (default: "")
* `collector_vmstat_fields` Regexp of fields to return for vmstat collector. (default: "^(oom\_kill|pgpg|pswp|pg.*fault).*")
* `collector_wifi_fixtures` Test fixtures to use for wifi collector metrics.
* `collectors_disabled` An array of explicitly disabled collectors.
* `collectors_enabled` An array of explicitly enabled collectors.
* `custom_options` Use for your configuration if defined proterties are not satisfying your needs.
* `log_format` Where to send log files. (default: "logger:stdout")
* `log_level` Only log messages with the given severity or above. Valid levels: [debug, info, warn, error, fatal]
* `path_procfs` procfs mountpoint. (default: "/proc")
* `path_rootfs` rootfs mountpoint. (default: "/")
* `path_sysfs` sysfs mountpoint. (default: "/sys")
* `user` System user to run node exporter as. (default "root") **change this to a non-root user if possible**
* `web_disable_exporter_metrics` Exclude metrics about the exporter itself. (promhttp_*, process_*, go_*)
* `web_listen_address` Address to listen on for web interface and telemetry. (default: ":9100")
* `web_max_requests` Maximum number of parallel scrape requests. Use 0 to disable. (default: "40")
* `web_telemetry_path` Path under which to expose metrics. (default: "/metrics")

```ruby

listen_ip = '127.0.0.1'

node_exporter 'main' do
  web_listen_address "#{listen_ip}:9100"
  action [:enable, :start]
end
```

or just set

* `node['prometheus_exporters']['listen_interface']`
* `node['prometheus_exporters']['node']['collectors']`
* `node['prometheus_exporters']['node']['textfile_directory']`
* `node['prometheus_exporters']['node']['ignored_net_devs']`
* `node['prometheus_exporters']['node']['user']`

and add `recipe['prometheus_exporters::node]` to your run_list.

## postgres_exporter

The postgres_exporter resource supports running multiple copies of PostgreSQL exporter the same system. This is useful if you have multiple copies of PostgreSQL running on the same system
(eg. different versions) or you are connecting to multiple remote PostgreSQL servers across the network.

* `constant_labels` A list of label=value separated by comma(,).
* `data_source_name` PostgreSQL connection string. E.g. `postgresql://login:password@hostname:port/dbname`
* `data_source_pass_file` The same as above but reads the password from a file.

CHANGELOG.md
# prometheus\_exporters

## 0.17.2

- [Kirill Kuznetsov] Blackbox exporter 0.17.0

## 0.17.1

- [Kirill Kuznetsov] Fix cookstyle offences
- [Kirill Kuznetsov] Bundle update

## 0.17.0

- [Shai Atias] apache exporter 0.8.0
- [Kirill Kuznetsov] fixed bug with `--insecure` arg for apache exporter
- [Kirill Kuznetsov] mongodb exporter 0.11.0
- [Kirill Kuznetsov] node exporter 1.0.0
- [Kirill Kuznetsov] process exporter 0.6.0
- [Kirill Kuznetsov] redis exporter 1.6.1
- [Kirill Kuznetsov] snmp exporter 0.18.0
- [Kirill Kuznetsov] statsd exporter 0.16.0
- [Kirill Kuznetsov] wmi exporter 0.13.0
- [Kirill Kuznetsov] node exporter resource attributes update

## 0.16.3

- [Kirill Kuznetsov] set consul exporter health summary property to false by default

## 0.16.2

- [Sergei Makarevich] consul exporter 0.6.0
- [Kirill Kuznetsov] consul exporter resource refacotring
- [Kirill Kuznetsov] Ubuntu 20.04 test suites

## 0.16.1

- [Lior Tzur] rabbitmq_exporter 1.0.0-RC7
- [Lior Tzur] recipe[prometheus_exporters::rabbitmq] default publish address: ''

## 0.16.0

- [Lior Tzur] switch the rabbitmq expoerter back to ENV-var style configuration
- [Kirill Kuznetsov] update chefspec tests
- [Kirill Kuznetsov] add tests that rabbitmq accepts its configuration
- [Kirill Kuznetsov] TravisCI was switched to ruby 2.7.1

## 0.15.9

- [Kirill Kuznetsov] fix cookstyle bugs
- [Kirill Kuznetsov] fix rabbitmq systemd unit env variable

## 0.15.8

- [Lior Tzur] rabbitmq_exporter
- [Kirill Kuznetsov] rabbitmq_exporter resource refacotring
- [Kirill Kuznetsov] bundler update
- [Kirill Kuznetsov] testrig cookbook refacotring

## 0.15.7

- [Adam Kobi] elasticsearch_exporter
- [Adam Kobi] mongodb_exporter
- [Kirill Kuznetsov] README update & alphabetical sort
- [Kirill Kuznetsov] Chef 13 tests were dropped
- [Kirill Kuznetsov] Fedora 30/31 tests only
- [Kirill Kuznetsov] Ubuntu 14 support was dropped

## 0.15.6

- [Kirill Kuznetsov] varnish recipe fix

## 0.15.5

- [Stephan Scheying] initial varnish exporter support
- [Kirill Kuznetsov] full support for cmd line parameters for varnish exporter.

## 0.15.4

- [Kirill Kuznetsov] one more bundle update
- [Kirill Kuznetsov] fix travisci.yaml bundle preparation
- [Wolfgang Schnerring] postgres\_exporter `disable\_settings\_metrics` option support

## 0.15.3

- [Kirill Kuznetsov] - bundle update

## 0.15.2

- [Kirill Kuznetsov] - exporters update:
  - blackbox\_exporter - `0.16.0`
  - postgres\_exporter - `0.8.0`
  - redis\_exporter - `1.3.5`
  - snmp\_exporter - `0.16.1`
  - statsd\_exporter - `0.13.0`
  - wmi\_exporter - `0.9.0`
- [Kirill Kuznetsov] - TravisCI ruby 2.6.5

## 0.15.1

- [Kirill Kuznetsov] - dummy release to test TravisCI flow.

## 0.15.0

- [Kirill Kuznetsov] - exporters update:
  - apache\_exporter - `0.7.0`
  - blackbox\_exporter - `0.15.0`
  - mysqld\_exporter - `0.12.1`
  - node\_exporter - `0.18.1`
  - postgres\_exporter - `0.5.1`
  - redis\_exporter - `1.1.1`
  - statsd\_exporter - `0.12.2`
  - wmi\_exporter - `0.8.3`
- [Kirill Kuznetsov] - Test Kitchen & TravisCI now use Dokken for tests
- [Kirill Kuznetsov] - bundle update

## 0.14.1

- [Asher Yanich] Fix bad quoting on systemd\_unit\_whitelist and systemd\_unit\_blacklist resources.

## 0.14.0

*Breaking changes*:

If you want to use Redis exporter < `1.0.0` consider a version of this cookbook lower than `0.14.0`.

- [Brian Baker] - Redis exporter 1.0.0 installation support.
- [Kirill Kuznetsov] - Redis exporter 1.0.0 full support.
- [Kirill Kuznetsov] - exporters update:
  - redis\_exporter - `1.0.0`

## 0.13.1

- [Wolfgang Schnerring] - statsd\_exporter resource.

## 0.13.0

- [Kirill Kuznetsov] - fix #23
- [Kirill Kuznetsov] - exporters update:
  - blackbox\_exporter - `0.14.0`
  - haproxy\_exporter - `0.10.0`
  - node\_exporter - `0.18.0`
  - process\_exporter - `0.5.0`
  - redis\_exporter - `0.34.1`
  - snmp\_exporter - `0.15.0`
  - wmi\_exporter - `0.7.0`
- [Kirill Kuznetsov] - actualize properties for all resources and refactor their names a bit.
- [Kirill Kuznetsov] - Chef 15 tests and support.
- [Kirill Kuznetsov] - TravisCI ruby 2.6.3 & matrix update.

## 0.12.1

- [Fatih Sarhan] - add Apache exporter to resource list in README

## 0.12.0

- [Fatih Sarhan] - resource list for README
- [Fatih Sarhan] - Apache exporter resource

## 0.11.1

- [Charles Rowe] - fix publication to supermarket.

## 0.11.0

- [Charles Rowe] - Add haproxy\_exporter

## 0.10.1

- [Kirill Kuznetsov] - fix mysqld\_exporter cmd for older systemd versions.
- [Kirill Kuznetsov] - add chefspec tests for `cookbook[testrig]`

## 0.10.0

- [Kirill Kuznetsov] - exporters update:
  - mysqld\_exporter: `0.11.0`
  - node\_exporter: `0.17.0`
  - postgres\_exporter: `0.4.7`
  - redis\_exporter: `0.22.1`
  - snmp\_exporter: `0.13.0`
  - wmi\_exporter: `0.5.0`

## 0.9.0

- [Viktor Radnai] - Add process\_exporter

## 0.8.3

- [Edwin Mourant] - Smaill fix to initscript template

## 0.8.2

- [Viktor Radnai] - blackbox\_exporter readme.

## 0.8.1

- [Kirill Kuznetsov] CHANGELOG update
- [Kirill Kuznetsov] RuboCop offences fix
- [Kirill Kuznetsov] Add 'enabled' attribute for wmi\_exporter

## 0.8.0

- [Yousef Alam] - Add blackbox\_exporter
- [Viktor Radnai] - Small fix to postgres\_exporter binary's symlink


## 0.7.0

- [Kieren Scott] - Add mysqld\_exporter
- [Matt Mencel] - Add wmi\_exporter
- [Viktor Radnai] - Add 'enabled' attribute for using with Chef search
- [Viktor Radnai] - Fix: Remove unnecessary install step from redis exporter start action
- [Viktor Radnai] - Fix: quote environment variable values in init script

## 0.6.1

- [Kirill Kuznetsov] - fixed #7
- [Kirill Kuznetsov] - fixed exporter services naming; it's possible to install more than one copy of an exporter with a different service name
- [Denis C.] - node\_exporter port attribute for `recipe[prometheus_exporters::node]`

## 0.6.0

- [Kirill Kuznetsov] - Chef 14 support was tested
- [Kirill Kuznetsov] - Ubuntu 18.04 support was tested

resources/node.rb
#
# Cookbook:: prometheus_exporters
# Resource:: node
#
# Copyright:: 2017, Evil Martians
#
# All rights reserved - Do Not Redistribute
#

COLLECTOR_LIST = %w(
  arp
  bcache
  bonding
  buddyinfo
  conntrack
  cpu
  cpufreq
  diskstats
  drbd
  edac
  entropy
  filefd
  filesystem
  hwmon
  infiniband
  interrupts
  ipvs
  ksmd
  loadavg
  logind
  mdadm
  meminfo
  meminfo_numa
  mountstats
  netclass
  netdev
  netstat
  nfs
  nfsd
  ntp
  perf
  pressure
  processes
  qdisc
  runit
  sockstat
  stat
  supervisord
  systemd
  tcpstat
  textfile
  time
  timex
  uname
  vmstat
  wifi
  xfs
  zfs
).freeze

resource_name :node_exporter
provides :node_exporter

property :collectors_enabled, Array, callbacks: {
  'should be a collector' => lambda do |collectors|
    collectors.all? { |element| COLLECTOR_LIST.include? element }
  end,
}, default: node['prometheus_exporters']['node']['collectors_enabled'].dup
property :collectors_disabled, Array, callbacks: {
  'should be a collector' => lambda do |collectors|
    collectors.all? { |element| COLLECTOR_LIST.include? element }
  end,
}, default: []
property :collector_diskstats_ignored_devices, String
property :collector_filesystem_ignored_fs_types, String
property :collector_filesystem_ignored_mount_points, String
property :collector_netclass_ignored_devices, String
property :collector_netdev_device_blacklist, String
property :collector_netdev_device_whitelist, String
property :collector_netstat_fields, String
property :collector_ntp_ip_ttl, [String, Integer]
property :collector_ntp_local_offset_tolerance, String
property :collector_ntp_max_distance, String
property :collector_ntp_protocol_version, [String, Integer]
property :collector_ntp_server, String
property :collector_ntp_server_is_local, [true, false]
property :collector_qdisc_fixtures, String
property :collector_perf_cpus, String
property :collector_runit_servicedir, String
property :collector_supervisord_url, String
property :collector_systemd_enable_restarts_metrics, [true, false], default: false
property :collector_systemd_enable_start_time_metrics, [true, false], default: false
property :collector_systemd_enable_task_metrics, [true, false], default: false
property :collector_systemd_private, [true, false], default: false
property :collector_systemd_unit_blacklist, String
property :collector_systemd_unit_whitelist, String
property :collector_textfile_directory, String
property :collector_vmstat_fields, String
property :collector_wifi_fixtures, String
property :log_format, String, default: 'logfmt'
property :log_level, String, default: 'info'
property :path_procfs, String
property :path_rootfs, String
property :path_sysfs, String
property :user, String, default: 'root'
property :web_disable_exporter_metrics, [true, false], default: false
property :web_listen_address, String, default: ':9100'
property :web_max_requests, [String, Integer], default: 40
property :web_telemetry_path, String, default: '/metrics'

property :custom_options, String

action :install do
  # Set property that can be queried with Chef search
  node.default['prometheus_exporters']['node']['enabled'] = true

  options = "--web.listen-address=#{new_resource.web_listen_address}"
  options += ' --collector.ntp.server-is-local' if new_resource.collector_ntp_server_is_local
  options += ' --collector.systemd.enable-restarts-metrics' if new_resource.collector_systemd_enable_restarts_metrics
  options += ' --collector.systemd.enable-start-time-metrics' if new_resource.collector_systemd_enable_start_time_metrics
  options += ' --collector.systemd.enable-task-metrics' if new_resource.collector_systemd_enable_task_metrics
  options += ' --collector.systemd.private' if new_resource.collector_systemd_private
  options += ' --web.disable-exporter-metrics' if new_resource.web_disable_exporter_metrics
  options += " --collector.diskstats.ignored-devices='#{new_resource.collector_diskstats_ignored_devices}'" if new_resource.collector_diskstats_ignored_devices
  options += " --collector.filesystem.ignored-fs-types='#{new_resource.collector_filesystem_ignored_fs_types}'" if new_resource.collector_filesystem_ignored_fs_types
  options += " --collector.filesystem.ignored-mount-points='#{new_resource.collector_filesystem_ignored_mount_points}'" if new_resource.collector_filesystem_ignored_mount_points
  options += " --collector.netclass.ignored-devices='#{new_resource.collector_netclass_ignored_devices}'" if new_resource.collector_netclass_ignored_devices
  options += " --collector.netdev.device-blacklist='#{new_resource.collector_netdev_device_blacklist}'" if new_resource.collector_netdev_device_blacklist
  options += " --collector.netdev.device-whitelist='#{new_resource.collector_netdev_device_whitelist}'" if new_resource.collector_netdev_device_whitelist
  options += " --collector.netstat.fields='#{new_resource.collector_netstat_fields}'" if new_resource.collector_netstat_fields
  options += " --collector.ntp.ip-ttl=#{new_resource.collector_ntp_ip_ttl}" if new_resource.collector_ntp_ip_ttl
  options += " --collector.ntp.local-offset-tolerance=#{new_resource.collector_ntp_local_offset_tolerance}" if new_resource.collector_ntp_local_offset_tolerance
  options += " --collector.ntp.max-distance=#{new_resource.collector_ntp_max_distance}" if new_resource.collector_ntp_max_distance
  options += " --collector.ntp.protocol-version=#{new_resource.collector_ntp_protocol_version}" if new_resource.collector_ntp_protocol_version
  options += " --collector.ntp.server=#{new_resource.collector_ntp_server}" if new_resource.collector_ntp_server
  options += " --collector.qdisc.fixtures='#{new_resource.collector_qdisc_fixtures}'" if new_resource.collector_qdisc_fixtures
  options += " --collector.perf.cpus='#{new_resource.collector_perf_cpus}'" if new_resource.collector_perf_cpus
  options += " --collector.runit.servicedir='#{new_resource.collector_runit_servicedir}'" if new_resource.collector_runit_servicedir
  options += " --collector.supervisord.url='#{new_resource.collector_supervisord_url}'" if new_resource.collector_supervisord_url
  options += " --collector.systemd.unit-blacklist=#{new_resource.collector_systemd_unit_blacklist}" if new_resource.collector_systemd_unit_blacklist
  options += " --collector.systemd.unit-whitelist=#{new_resource.collector_systemd_unit_whitelist}" if new_resource.collector_systemd_unit_whitelist
  options += " --collector.textfile.directory=#{new_resource.collector_textfile_directory}" if new_resource.collector_textfile_directory
  options += " --collector.vmstat.fields='#{new_resource.collector_vmstat_fields}'" if new_resource.collector_vmstat_fields
  options += " --collector.wifi.fixtures='#{new_resource.collector_wifi_fixtures}'" if new_resource.collector_wifi_fixtures
  options += " --log.format=#{new_resource.log_format}"
  options += " --log.level=#{new_resource.log_level}"
  options += " --path.procfs=#{new_resource.path_procfs}" if new_resource.path_procfs
  options += " --path.rootfs=#{new_resource.path_rootfs}" if new_resource.path_rootfs
  options += " --path.sysfs=#{new_resource.path_sysfs}" if new_resource.path_sysfs
  options += " --web.max-requests=#{new_resource.web_max_requests}"
  options += " --web.telemetry-path=#{new_resource.web_telemetry_path}"

  options += " #{new_resource.custom_options}" if new_resource.custom_options

  options += new_resource.collectors_enabled.map { |c| " --collector.#{c}" }.join
  options += new_resource.collectors_disabled.map { |c| " --no-collector.#{c}" }.join unless new_resource.collectors_disabled.empty?

  service_name = "node_exporter_#{new_resource.name}"

  # Download binary
  remote_file 'node_exporter' do
    path "#{Chef::Config[:file_cache_path]}/node_exporter.tar.gz"
    owner 'root'
    group 'root'
    mode '0644'
    source node['prometheus_exporters']['node']['url']
    checksum node['prometheus_exporters']['node']['checksum']
    notifies :restart, "service[#{service_name}]"
  end

  bash 'untar node_exporter' do
    code "tar -xzf #{Chef::Config[:file_cache_path]}/node_exporter.tar.gz -C /opt"
    action :nothing
    subscribes :run, 'remote_file[node_exporter]', :immediately
  end

  link '/usr/local/sbin/node_exporter' do
    to "/opt/node_exporter-#{node['prometheus_exporters']['node']['version']}.linux-amd64/node_exporter"
  end

  # Configure to run as a service
  service service_name do
    action :nothing
  end

  case node['init_package']
  when /init/
    %w(
      /var/run/prometheus
      /var/log/prometheus
    ).each do |dir|
      directory dir do
        owner 'root'
        group 'root'
        mode '0755'
        recursive true
        action :create
      end
    end

    directory "/var/log/prometheus/#{service_name}" do
      owner new_resource.user
      group 'root'
      mode '0755'
      action :create
    end

    template "/etc/init.d/#{service_name}" do
      cookbook 'prometheus_exporters'
      source 'initscript.erb'
      owner 'root'
      group 'root'
      mode '0755'
      variables(
        name: service_name,
        user: new_resource.user,
        cmd: "/usr/local/sbin/node_exporter #{options}",
        service_description: 'Prometheus Node Exporter',
      )
      notifies :restart, "service[#{service_name}]"
    end

  when /systemd/
    systemd_unit "#{service_name}.service" do
      content(
        'Unit' => {
          'Description' => 'Systemd unit for Prometheus Node Exporter',
          'After' => 'network.target remote-fs.target apiserver.service',
        },
        'Service' => {
          'Type' => 'simple',
          'User' => new_resource.user,
          'ExecStart' => "/usr/local/sbin/node_exporter #{options}",
          'WorkingDirectory' => '/',
          'Restart' => 'on-failure',
          'RestartSec' => '30s',
        },
        'Install' => {
          'WantedBy' => 'multi-user.target',
        },
      )
      notifies :restart, "service[#{service_name}]"
      action :create
    end

  when /upstart/
    template "/etc/init/#{service_name}.conf" do
      cookbook 'prometheus_exporters'
      source 'upstart.conf.erb'
      owner 'root'
      group 'root'
      mode '0644'
      variables(
        cmd: "/usr/local/sbin/node_exporter #{options}",
        user: new_resource.user,
        service_description: 'Prometheus Node Exporter',
      )
      notifies :restart, "service[#{service_name}]"
    end

  else
    raise "Init system '#{node['init_package']}' is not supported by the 'prometheus_exporters' cookbook"
  end

  if new_resource.collector_textfile_directory and
     new_resource.collector_textfile_directory != ''
    directory 'collector_textfile_directory' do
      path new_resource.collector_textfile_directory
      owner 'root'
      group 'root'
      mode '0755'
      action :create
      recursive true
    end
  end
end

action :enable do
  action_install
  service "node_exporter_#{new_resource.name}" do
    action :enable
  end
end

action :start do
  service "node_exporter_#{new_resource.name}" do
    action :start
  end
end

action :disable do
  service "node_exporter_#{new_resource.name}" do
    action :disable
  end
end

action :stop do
  service "node_exporter_#{new_resource.name}" do
    action :stop
  end
end
